### ISSUE TITLE
thread_local in enclave

### ISSUE THREAD

**Issue by insujang** [intel/linux-sgx#283]
Dear,

I have a problem about thread-local storage in enclave.

According to [the Intel SGX Developer Reference for Linux document](https://download.01.org/intel-sgx/linux-2.1.3/docs/Intel_SGX_Developer_Reference_Linux_2.1.3_Open_Source.pdf), Intel SGX supports c++11 standard thread-local storage, i.e. `thread_local`.

However, it seems not work if a thread_local variable is a global variable.

trusted/sample.cpp
```
thread_local int a;
void ecall_num1 () {
    a = 4;
    ocall_printf("a: %d from ecall_num1\n", a);
}
void ecall_num2 () {
    ocall_printf("a: %d from ecall_num2\n", a);
}
```

untrusted/sample.cpp
```
// initialize enclave
ecall_num1(enclave_id);
ecall_num2(enclave_id);
```

Expected output:
```
a: 4 from ecall_num1
a: 4 from ecall_num2
```
Actual output:
```
a: 4 from ecall_num1
a: 0 from ecall_num2
```

It is an intended output? Why the value of thread_local global variable is erased?

**Comment by insujang** [comment]
Additional note: enclave under simulation mode the output is:
```
a: 4 from ecall_num1
a: 4 from ecall_num2
```
Only under HW mode the second ecall prints a=0.

**Comment by MWShan** [comment]
What TCSPolicy have you selected in the configuration file?  If you select a TCSPolicy of 0, then the trusted thread (TCS) is not bound to the untrusted thread.  In this case the thread local storage data is reinitialized on each root ecall.  Can you set the TCSPolicy to 1 and retry your test?

**Comment by insujang** [comment]
@MWShan Thanks for the reply!
I see the configuration xml file and TCSPolicy is set to 1 by default and thread_local storage is  
reinitialized for each ecall.
Surprisingly, if I set TCSPolicy to 0, this thread local storage is not reinitialized.

I find the corresponding part from the Intel SGX development reference and it says
> TCSPolicy
> Value 0 – TCS is bound to the untrusted thread.
> Value 1 – TCS is not bound to the untrusted thread.

So in my case I should have set it to 0.

I didn't know about TCSPolicy but I got it now :)
Thank you again for your help!
